pennylane/_device.py

```diff
@@ -216,12 +216,22 @@ def execute(self, queue, observables):
                     results.append(self.expval(obs.name, obs.wires, obs.parameters))
                 elif obs.return_type == "variance":
                     results.append(self.var(obs.name, obs.wires, obs.parameters))
+                elif obs.return_type == "sample":
+                    if not hasattr(obs, 'num_samples'):
+                        raise DeviceError("Number of samples not specified for observable {}".format(obs.name))
+
+                    results.append(self.sample(obs.name, obs.wires, obs.parameters, obs.num_samples))
```

Comment:
Setting the `Operation` attribute `return_type` as a string felt a bit hacky to me in #232, but it was a stop-gap approach to avoid making even larger changes during that refactor. But nice that you can add others, such as `obs.num_samples`!

Perhaps in another PR, it would be 'cleaner' to use a [named tuple](https://docs.python.org/3/library/collections.html#collections.namedtuple) or an [enumeration class](https://docs.python.org/3/library/enum.html) for this attributes which can only take certain values?

Reply:
I'm always in favour of enums! But as python is an interpreted language they are not as useful as, say, in C++.

Actually as I'm programming more and more in Python, I feel like 'being hacky' is a design principle of this language.

Reply:
I'd keep this for the future though when we refactor the QNode.